File: agent-worker.js
/**
 * Agent Worker Module
 *
 * Provides agent worker thread pool with management, task queue, and concurrency.
 * Usage: node agent-worker.js [command] [options]
 *
 * Commands:
 *   demo                    Run demo
 *   submit <task>         Submit a task
 *   status                 Show worker status
 */

const fs = require('fs');
const path = require('path');

const DATA_DIR = path.join(__dirname, 'data');
const WORKER_DB = path.join(DATA_DIR, 'worker-state.json');

function ensureDataDir() {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
}

function loadJSON(file, defaultVal = {}) {
  ensureDataDir();
  if (!fs.existsSync(file)) {
    return defaultVal;
  }
  try {
    return JSON.parse(fs.readFileSync(file, 'utf-8'));
  } catch {
    return defaultVal;
  }
}

function saveJSON(file, data) {
  ensureDataDir();
  fs.writeFileSync(file, JSON.stringify(data, null, 2));
}

/**
 * Worker State
 */
const WorkerState = {
  IDLE: 'idle',
  BUSY: 'busy',
  TERMINATED: 'terminated'
};

/**
 * Task
 */
class Task {
  constructor(id, handler, options = {}) {
    this.id = id;
    this.handler = handler;
    this.options = {
      timeout: options.timeout || 30000,
      priority: options.priority || 0,
      retries: options.retries || 0,
      ...options
    };
    this.state = 'pending'; // pending, running, completed, failed, cancelled
    this.result = null;
    this.error = null;
    this.startTime = null;
    this.endTime = null;
    this.attempts = 0;
  }

  async execute() {
    this.state = 'running';
    this.startTime = Date.now();
    this.attempts++;

    try {
      const result = await this.runWithTimeout();
      this.state = 'completed';
      this.result = result;
      this.endTime = Date.now();
      return { success: true, result };
    } catch (error) {
      this.state = 'failed';
      this.error = error.message;
      this.endTime = Date.now();
      return { success: false, error: error.message };
    }
  }

  runWithTimeout() {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error(`Task timeout after ${this.options.timeout}ms`));
      }, this.options.timeout);

      Promise.resolve(this.handler()).then(result => {
        clearTimeout(timer);
        resolve(result);
      }).catch(err => {
        clearTimeout(timer);
        reject(err);
      });
    });
  }

  getDuration() {
    if (this.startTime && this.endTime) {
      return this.endTime - this.startTime;
    }
    return 0;
  }
}

/**
 * Worker
 */
class Worker {
  constructor(id, options = {}) {
    this.id = id;
    this.state = WorkerState.IDLE;
    this.currentTask = null;
    this.options = {
      maxConcurrent: options.maxConcurrent || 1,
      timeout: options.timeout || 60000,
      ...options
    };
    this.stats = {
      tasksProcessed: 0,
      tasksSucceeded: 0,
      tasksFailed: 0,
      totalRuntime: 0,
      lastTask: null
    };
  }

  async executeTask(task) {
    this.state = WorkerState.BUSY;
    this.currentTask = task;

    const startTime = Date.now();
    const result = await task.execute();
    const duration = Date.now() - startTime;

    this.stats.tasksProcessed++;
    this.stats.totalRuntime += duration;
    this.stats.lastTask = task.id;

    if (result.success) {
      this.stats.tasksSucceeded++;
    } else {
      this.stats.tasksFailed++;
    }

    this.state = WorkerState.IDLE;
    this.currentTask = null;

    return result;
  }

  isAvailable() {
    return this.state === WorkerState.IDLE;
  }

  getStats() {
    return {
      id: this.id,
      state: this.state,
      currentTask: this.currentTask?.id || null,
      ...this.stats
    };
  }

  terminate() {
    this.state = WorkerState.TERMINATED;
  }
}

/**
 * Task Queue
 */
class TaskQueue {
  constructor(options = {}) {
    this.options = {
      maxSize: options.maxSize || 1000,
      priority: options.priority || true,
      ...options
    };
    this.tasks = [];
    this.waiting = [];
  }

  enqueue(task) {
    if (this.tasks.length >= this.options.maxSize) {
      return { error: 'Queue full' };
    }

    this.tasks.push(task);
    this.tasks.sort((a, b) => b.options.priority - a.options.priority);

    return { success: true, taskId: task.id };
  }

  dequeue() {
    return this.tasks.shift() || null;
  }

  peek() {
    return this.tasks[0] || null;
  }

  size() {
    return this.tasks.length;
  }

  clear() {
    this.tasks = [];
  }
}

/**
 * Worker Pool
 */
class WorkerPool {
  constructor(options = {}) {
    this.options = {
      minWorkers: options.minWorkers || 2,
      maxWorkers: options.maxWorkers || 10,
      maxQueueSize: options.maxQueueSize || 1000,
      workerTimeout: options.workerTimeout || 60000,
      ...options
    };
    this.workers = new Map();
    this.queue = new TaskQueue({ maxSize: this.options.maxQueueSize });
    this.running = false;
    this.stats = {
      tasksSubmitted: 0,
      tasksCompleted: 0,
      tasksFailed: 0
    };
  }

  // Initialize workers
  initialize() {
    for (let i = 0; i < this.options.minWorkers; i++) {
      const workerId = `worker-${i}`;
      this.workers.set(workerId, new Worker(workerId));
    }
  }

  // Get available worker
  getAvailableWorker() {
    for (const worker of this.workers.values()) {
      if (worker.isAvailable()) {
        return worker;
      }
    }
    return null;
  }

  // Scale workers
  scaleUp(count = 1) {
    const currentCount = this.workers.size;
    const newCount = Math.min(currentCount + count, this.options.maxWorkers);

    for (let i = currentCount; i < newCount; i++) {
      const workerId = `worker-${i}`;
      this.workers.set(workerId, new Worker(workerId));
    }

    return { workers: newCount };
  }

  scaleDown(count = 1) {
    const currentCount = this.workers.size;
    const newCount = Math.max(currentCount - count, this.options.minWorkers);

    // Remove idle workers
    let removed = 0;
    for (const [id, worker] of this.workers) {
      if (removed >= count) break;
      if (worker.isAvailable()) {
        worker.terminate();
        this.workers.delete(id);
        removed++;
      }
    }

    return { workers: this.workers.size };
  }

  // Submit task
  submit(handler, options = {}) {
    const taskId = `task-${Date.now()}-${Math.random().toString(36).substr(2, 4)}`;
    const task = new Task(taskId, handler, options);

    // Try to assign to worker
    const worker = this.getAvailableWorker();

    if (worker) {
      this.stats.tasksSubmitted++;
      worker.executeTask(task).then(result => {
        if (result.success) {
          this.stats.tasksCompleted++;
        } else {
          this.stats.tasksFailed++;
        }
      });
      return { assigned: true, taskId, workerId: worker.id };
    } else {
      // Add to queue
      this.queue.enqueue(task);
      this.stats.tasksSubmitted++;
      return { queued: true, taskId, queueSize: this.queue.size() };
    }
  }

  // Process queue
  async processQueue() {
    while (this.queue.size() > 0) {
      const worker = this.getAvailableWorker();
      if (!worker) break;

      const task = this.queue.dequeue();
      if (!task) break;

      worker.executeTask(task).then(result => {
        if (result.success) {
          this.stats.tasksCompleted++;
        } else {
          this.stats.tasksFailed++;
        }
      });
    }
  }

  // Get worker stats
  getWorkerStats(workerId) {
    const worker = this.workers.get(workerId);
    return worker ? worker.getStats() : null;
  }

  // Get all workers
  getAllWorkers() {
    return Array.from(this.workers.values()).map(w => w.getStats());
  }

  // Get queue size
  getQueueSize() {
    return this.queue.size();
  }

  // Get stats
  getStats() {
    const idleWorkers = Array.from(this.workers.values()).filter(w => w.isAvailable()).length;
    const busyWorkers = this.workers.size - idleWorkers;

    return {
      workers: this.workers.size,
      idle: idleWorkers,
      busy: busyWorkers,
      queueSize: this.queue.size(),
      maxWorkers: this.options.maxWorkers,
      tasks: {
        submitted: this.stats.tasksSubmitted,
        completed: this.stats.tasksCompleted,
        failed: this.stats.tasksFailed
      }
    };
  }
}

/**
 * Demo
 */
async function demo() {
  console.log('=== Agent Worker Demo ===\n');

  const pool = new WorkerPool({ minWorkers: 2, maxWorkers: 5 });
  pool.initialize();

  // Show initial state
  console.log('1. Initial Workers:');
  const initialStats = pool.getStats();
  console.log(`   Workers: ${initialStats.workers}`);
  console.log(`   Idle: ${initialStats.idle}`);

  // Submit tasks
  console.log('\n2. Submitting Tasks:');

  for (let i = 1; i <= 5; i++) {
    const result = pool.submit(async () => {
      await new Promise(r => setTimeout(r, 50));
      return { taskId: i, result: i * 10 };
    }, { priority: 5 - i });
    console.log(`   Task ${i}: ${result.assigned ? `assigned to ${result.workerId}` : 'queued'}`);
  }

  // Wait for tasks to complete
  await new Promise(r => setTimeout(r, 200));

  // Show stats
  console.log('\n3. Task Statistics:');
  const stats = pool.getStats();
  console.log(`   Submitted: ${stats.tasks.submitted}`);
  console.log(`   Completed: ${stats.tasks.completed}`);
  console.log(`   Failed: ${stats.tasks.failed}`);

  // Show worker stats
  console.log('\n4. Worker Statistics:');
  const workers = pool.getAllWorkers();
  workers.forEach(w => {
    console.log(`   ${w.id}: state=${w.state}, processed=${w.tasksProcessed}, succeeded=${w.tasksSucceeded}`);
  });

  // Scale up
  console.log('\n5. Scaling Up:');
  pool.scaleUp(2);
  const afterScale = pool.getStats();
  console.log(`   Workers: ${afterScale.workers}`);

  // Scale down
  console.log('\n6. Scaling Down:');
  pool.scaleDown(1);
  const afterScaleDown = pool.getStats();
  console.log(`   Workers: ${afterScaleDown.workers}`);

  // Submit more tasks with different priorities
  console.log('\n7. Priority Tasks:');
  pool.submit(async () => 'low priority', { priority: 1 });
  pool.submit(async () => 'high priority', { priority: 10 });
  console.log('   Submitted: low priority (1), high priority (10)');

  await new Promise(r => setTimeout(r, 100));

  // Final stats
  console.log('\n8. Final Statistics:');
  const finalStats = pool.getStats();
  console.log(`   Workers: ${finalStats.workers} (idle: ${finalStats.idle}, busy: ${finalStats.busy})`);
  console.log(`   Queue: ${finalStats.queueSize}`);
  console.log(`   Tasks: ${finalStats.tasks.submitted} submitted, ${finalStats.tasks.completed} completed`);

  // Show queue info
  console.log('\n9. Queue Management:');
  console.log(`   Current queue size: ${pool.getQueueSize()}`);

  console.log('\n=== Demo Complete ===');
}

// CLI
const args = process.argv.slice(2);
const cmd = args[0];

if (cmd === 'demo') {
  demo();
} else if (cmd === 'submit') {
  const pool = new WorkerPool();
  pool.initialize();
  pool.submit(async () => 'done').then(r => console.log(JSON.stringify(r, null, 2)));
} else if (cmd === 'status') {
  const pool = new WorkerPool();
  console.log(JSON.stringify(pool.getStats(), null, 2));
} else {
  console.log('Agent Worker Module');
  console.log('Usage: node agent-worker.js [command]');
  console.log('Commands:');
  console.log('  demo              Run demo');
  console.log('  submit <task>  Submit a task');
  console.log('  status           Show worker status');
}

File: agent_worker.py
"""Agent Worker Module

Worker thread pool management for agent tasks including worker lifecycle,
task queue, concurrency control, worker scaling, and worker health monitoring.
"""
import time
import uuid
import threading
import asyncio
from typing import Dict, List, Optional, Any, Callable, Union
from dataclasses import dataclass, field
from enum import Enum
from collections import defaultdict, deque
import queue
import json


class WorkerState(str, Enum):
    """Worker states."""
    IDLE = "idle"
    RUNNING = "running"
    BUSY = "busy"
    STOPPING = "stopping"
    STOPPED = "stopped"
    ERROR = "error"


class TaskPriority(int, Enum):
    """Task priority levels."""
    LOW = 0
    NORMAL = 1
    HIGH = 2
    CRITICAL = 3


class WorkerStrategy(str, Enum):
    """Worker allocation strategies."""
    ROUND_ROBIN = "round_robin"
    LEAST_LOADED = "least_loaded"
    MOST_LOADED = "most_loaded"
    RANDOM = "random"
    PRIORITY = "priority"


@dataclass
class WorkerTask:
    """Worker task representation."""
    id: str
    func: Callable
    args: tuple = field(default_factory=tuple)
    kwargs: Dict[str, Any] = field(default_factory=dict)
    priority: TaskPriority = TaskPriority.NORMAL
    timeout: float = 30.0
    created_at: float = field(default_factory=time.time)
    started_at: Optional[float] = None
    completed_at: Optional[float] = None
    result: Any = None
    error: Optional[str] = None
    status: str = "pending"  # pending, running, completed, failed, timeout


@dataclass
class WorkerConfig:
    """Worker configuration."""
    name: str
    min_workers: int = 1
    max_workers: int = 10
    queue_size: int = 100
    task_timeout: float = 30.0
    idle_timeout: float = 300.0
    health_check_interval: float = 60.0
    auto_scale: bool = True
    scale_up_threshold: float = 0.8
    scale_down_threshold: float = 0.2
    strategy: WorkerStrategy = WorkerStrategy.LEAST_LOADED


@dataclass
class Worker:
    """Worker instance."""
    id: str
    name: str
    state: WorkerState = WorkerState.IDLE
    current_task: Optional[WorkerTask] = None
    tasks_completed: int = 0
    tasks_failed: int = 0
    total_runtime: float = 0.0
    created_at: float = field(default_factory=time.time)
    last_active: float = field(default_factory=time.time)
    thread: Optional[threading.Thread] = None


@dataclass
class WorkerStats:
    """Worker statistics."""
    total_tasks: int = 0
    completed_tasks: int = 0
    failed_tasks: int = 0
    timeout_tasks: int = 0
    avg_execution_time: float = 0.0
    total_runtime: float = 0.0
    peak_workers: int = 0


class WorkerPool:
    """Worker pool for managing worker threads."""

    def __init__(self, config: WorkerConfig):
        self.config = config
        self._lock = threading.RLock()
        self._workers: Dict[str, Worker] = {}
        self._task_queue: queue.PriorityQueue = queue.PriorityQueue(maxsize=config.queue_size)
        self._running = False
        self._stats = WorkerStats()
        self._hooks: Dict[str, List[Callable]] = defaultdict(list)
        self._task_results: Dict[str, Any] = {}
        self._worker_health: Dict[str, float] = {}

    def start(self):
        """Start the worker pool."""
        with self._lock:
            if self._running:
                return

            self._running = True

            # Start initial workers
            for i in range(self.config.min_workers):
                self._spawn_worker()

    def stop(self):
        """Stop the worker pool."""
        with self._lock:
            self._running = False

            # Signal all workers to stop
            for worker in self._workers.values():
                worker.state = WorkerState.STOPPING

            # Wait for workers to finish
            for worker in list(self._workers.values()):
                if worker.thread and worker.thread.is_alive():
                    worker.thread.join(timeout=5.0)

            self._workers.clear()

    def _spawn_worker(self) -> Worker:
        """Spawn a new worker."""
        worker_id = str(uuid.uuid4())[:8]
        worker_name = f"{self.config.name}-worker-{worker_id}"

        worker = Worker(
            id=worker_id,
            name=worker_name,
            state=WorkerState.IDLE
        )

        thread = threading.Thread(
            target=self._worker_loop,
            args=(worker,),
            daemon=True
        )
        thread.start()
        worker.thread = thread

        self._workers[worker_id] = worker
        self._worker_health[worker_id] = time.time()

        if len(self._workers) > self._stats.peak_workers:
            self._stats.peak_workers = len(self._workers)

        self._emit_event("worker_spawned", {"worker_id": worker_id, "name": worker_name})
        return worker

    def _worker_loop(self, worker: Worker):
        """Main worker loop."""
        while self._running and worker.state != WorkerState.STOPPING:
            try:
                # Get task from queue with timeout
                task = self._task_queue.get(timeout=1.0)

                # Execute task
                worker.state = WorkerState.BUSY
                worker.current_task = task
                task.status = "running"
                task.started_at = time.time()

                self._emit_event("task_started", {
                    "worker_id": worker.id,
                    "task_id": task.id
                })

                start_time = time.time()
                try:
                    # Execute with timeout
                    result = task.func(*task.args, **task.kwargs)
                    task.result = result
                    task.status = "completed"
                    task.completed_at = time.time()

                    worker.tasks_completed += 1
                    self._stats.completed_tasks += 1
                    self._task_results[task.id] = result

                except Exception as e:
                    task.error = str(e)
                    task.status = "failed"
                    task.completed_at = time.time()

                    worker.tasks_failed += 1
                    self._stats.failed_tasks += 1
                    self._task_results[task.id] = None

                    self._emit_event("task_failed", {
                        "worker_id": worker.id,
                        "task_id": task.id,
                        "error": str(e)
                    })

                execution_time = time.time() - start_time
                worker.total_runtime += execution_time
                self._stats.total_runtime += execution_time

                # Update stats
                if self._stats.completed_tasks > 0:
                    self._stats.avg_execution_time = (
                        self._stats.total_runtime / self._stats.completed_tasks
                    )

                worker.current_task = None
                worker.last_active = time.time()
                worker.state = WorkerState.IDLE

                self._task_queue.task_done()

                self._emit_event("task_completed", {
                    "worker_id": worker.id,
                    "task_id": task.id,
                    "execution_time": execution_time
                })

            except queue.Empty:
                # No task available, check idle timeout
                if time.time() - worker.last_active > self.config.idle_timeout:
                    if len(self._workers) > self.config.min_workers:
                        worker.state = WorkerState.STOPPING
                        break
                continue

            except Exception as e:
                worker.state = WorkerState.ERROR
                self._emit_event("worker_error", {
                    "worker_id": worker.id,
                    "error": str(e)
                })

        worker.state = WorkerState.STOPPED
        self._emit_event("worker_stopped", {"worker_id": worker.id})

    def submit_task(
        self,
        func: Callable,
        args: tuple = None,
        kwargs: Dict[str, Any] = None,
        priority: TaskPriority = TaskPriority.NORMAL,
        timeout: float = None
    ) -> str:
        """Submit a task to the worker pool."""
        task_id = str(uuid.uuid4())[:8]

        task = WorkerTask(
            id=task_id,
            func=func,
            args=args or (),
            kwargs=kwargs or {},
            priority=priority,
            timeout=timeout or self.config.task_timeout,
            created_at=time.time()
        )

        self._task_queue.put((priority.value, task))
        self._stats.total_tasks += 1

        self._emit_event("task_submitted", {
            "task_id": task_id,
            "priority": priority.name
        })

        # Auto-scale if enabled
        if self.config.auto_scale:
            self._check_scaling()

        return task_id

    def _check_scaling(self):
        """Check and perform auto-scaling."""
        queue_load = self._task_queue.qsize() / self.config.queue_size

        # Scale up
        if queue_load > self.config.scale_up_threshold:
            if len(self._workers) < self.config.max_workers:
                self._spawn_worker()

        # Scale down
        elif queue_load < self.config.scale_down_threshold:
            if len(self._workers) > self.config.min_workers:
                # Find idle workers to terminate
                for worker in self._workers.values():
                    if worker.state == WorkerState.IDLE:
                        worker.state = WorkerState.STOPPING
                        break

    def get_task_result(self, task_id: str, timeout: float = None) -> Any:
        """Get task result, waiting if necessary."""
        timeout = timeout or self.config.task_timeout
        start_time = time.time()

        while time.time() - start_time < timeout:
            if task_id in self._task_results:
                result = self._task_results.pop(task_id)
                return result
            time.sleep(0.1)

        return None

    def get_worker(self, worker_id: str) -> Optional[Worker]:
        """Get worker by ID."""
        with self._lock:
            return self._workers.get(worker_id)

    def list_workers(self, state: WorkerState = None) -> List[Worker]:
        """List workers, optionally filtered by state."""
        with self._lock:
            workers = list(self._workers.values())
            if state:
                workers = [w for w in workers if w.state == state]
            return workers

    def get_stats(self) -> Dict[str, Any]:
        """Get worker pool statistics."""
        with self._lock:
            return {
                "total_tasks": self._stats.total_tasks,
                "completed_tasks": self._stats.completed_tasks,
                "failed_tasks": self._stats.failed_tasks,
                "timeout_tasks": self._stats.timeout_tasks,
                "avg_execution_time": round(self._stats.avg_execution_time, 3),
                "total_runtime": round(self._stats.total_runtime, 3),
                "peak_workers": self._stats.peak_workers,
                "current_workers": len(self._workers),
                "queue_size": self._task_queue.qsize(),
                "queue_capacity": self.config.queue_size
            }

    def register_hook(self, event_type: str, handler: Callable):
        """Register an event hook."""
        self._hooks[event_type].append(handler)

    def _emit_event(self, event_type: str, data: Dict[str, Any]):
        """Emit a worker pool event."""
        for handler in self._hooks.get(event_type, []):
            try:
                handler(data)
            except Exception:
                pass


class AgentWorker:
    """Agent worker management system."""

    def __init__(self):
        self._lock = threading.RLock()
        self._pools: Dict[str, WorkerPool] = {}
        self._hooks: Dict[str, List[Callable]] = defaultdict(list)

    def create_pool(
        self,
        name: str,
        min_workers: int = 1,
        max_workers: int = 10,
        queue_size: int = 100,
        task_timeout: float = 30.0,
        idle_timeout: float = 300.0,
        health_check_interval: float = 60.0,
        auto_scale: bool = True,
        scale_up_threshold: float = 0.8,
        scale_down_threshold: float = 0.2,
        strategy: WorkerStrategy = WorkerStrategy.LEAST_LOADED
    ) -> str:
        """Create a new worker pool."""
        with self._lock:
            pool_id = str(uuid.uuid4())[:8]

            config = WorkerConfig(
                name=name,
                min_workers=min_workers,
                max_workers=max_workers,
                queue_size=queue_size,
                task_timeout=task_timeout,
                idle_timeout=idle_timeout,
                health_check_interval=health_check_interval,
                auto_scale=auto_scale,
                scale_up_threshold=scale_up_threshold,
                scale_down_threshold=scale_down_threshold,
                strategy=strategy
            )

            pool = WorkerPool(config)
            pool.start()

            self._pools[pool_id] = pool
            return pool_id

    def get_pool(self, pool_id: str) -> Optional[WorkerPool]:
        """Get pool by ID."""
        with self._lock:
            return self._pools.get(pool_id)

    def delete_pool(self, pool_id: str) -> bool:
        """Delete a worker pool."""
        with self._lock:
            pool = self._pools.get(pool_id)
            if not pool:
                return False

            pool.stop()
            del self._pools[pool_id]
            return True

    def list_pools(self) -> List[Dict[str, Any]]:
        """List all worker pools."""
        with self._lock:
            return [
                {
                    "id": pid,
                    "name": pool.config.name,
                    "min_workers": pool.config.min_workers,
                    "max_workers": pool.config.max_workers,
                    "current_workers": len(pool._workers),
                    "stats": pool.get_stats()
                }
                for pid, pool in self._pools.items()
            ]

    def submit_task(
        self,
        pool_id: str,
        func: Callable,
        args: tuple = None,
        kwargs: Dict[str, Any] = None,
        priority: TaskPriority = TaskPriority.NORMAL,
        timeout: float = None
    ) -> Optional[str]:
        """Submit a task to a pool."""
        pool = self.get_pool(pool_id)
        if not pool:
            return None

        return pool.submit_task(func, args, kwargs, priority, timeout)

    def get_task_result(self, pool_id: str, task_id: str, timeout: float = None) -> Any:
        """Get task result from a pool."""
        pool = self.get_pool(pool_id)
        if not pool:
            return None

        return pool.get_task_result(task_id, timeout)

    def get_pool_stats(self, pool_id: str) -> Optional[Dict[str, Any]]:
        """Get pool statistics."""
        pool = self.get_pool(pool_id)
        if not pool:
            return None

        return pool.get_stats()

    def get_pool_workers(self, pool_id: str) -> List[Dict[str, Any]]:
        """Get workers in a pool."""
        pool = self.get_pool(pool_id)
        if not pool:
            return []

        workers = pool.list_workers()
        return [
            {
                "id": w.id,
                "name": w.name,
                "state": w.state.value,
                "tasks_completed": w.tasks_completed,
                "tasks_failed": w.tasks_failed,
                "total_runtime": round(w.total_runtime, 3),
                "current_task_id": w.current_task.id if w.current_task else None
            }
            for w in workers
        ]

    def scale_pool(
        self,
        pool_id: str,
        min_workers: int = None,
        max_workers: int = None
    ) -> bool:
        """Scale a worker pool."""
        pool = self.get_pool(pool_id)
        if not pool:
            return False

        with pool._lock:
            if min_workers is not None:
                pool.config.min_workers = min_workers
            if max_workers is not None:
                pool.config.max_workers = max_workers

            # Adjust worker count
            current_count = len(pool._workers)
            if current_count < pool.config.min_workers:
                for _ in range(pool.config.min_workers - current_count):
                    pool._spawn_worker()
            elif current_count > pool.config.max_workers:
                # Mark excess workers for termination
                excess = current_count - pool.config.max_workers
                idle_workers = [w for w in pool._workers.values() if w.state == WorkerState.IDLE]
                for worker in idle_workers[:excess]:
                    worker.state = WorkerState.STOPPING

        return True

    def pause_pool(self, pool_id: str) -> bool:
        """Pause a worker pool (stop accepting new tasks)."""
        pool = self.get_pool(pool_id)
        if not pool:
            return False

        with pool._lock:
            for worker in pool._workers.values():
                worker.state = WorkerState.STOPPING

        return True

    def resume_pool(self, pool_id: str) -> bool:
        """Resume a paused worker pool."""
        pool = self.get_pool(pool_id)
        if not pool:
            return False

        with pool._lock:
            pool._running = True
            current_count = len(pool._workers)
            if current_count < pool.config.min_workers:
                for _ in range(pool.config.min_workers - current_count):
                    pool._spawn_worker()

        return True

    def register_hook(self, event_type: str, handler: Callable):
        """Register an event hook."""
        self._hooks[event_type].append(handler)

    def get_all_stats(self) -> Dict[str, Dict[str, Any]]:
        """Get statistics for all pools."""
        return {
            pid: pool.get_stats()
            for pid, pool in self._pools.items()
        }


# Global worker instance
agent_worker = AgentWorker()
